### PR TITLE
samples: nrf9160: gps: support circuitdojo_feather_nrf9160

### PR DESCRIPTION
**Pull request by mniestroj**
This platform has only external antenna, so narrow down possible Kconfig
options to reduce user confusion.

Set COEX0 pin to 1 when used with GPS frequencies, as the external
antenna signal goes through LNA, which has to be enabled.

MAPGPIOs are not used on this platform, so do not configure them.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

### CODE REVIEW

**Review by rlubos — APPROVED** [review]
LGTM

File: samples/nrf9160/gps/src/main.c [hunk 1]
```diff
@@ -40,6 +40,10 @@
 #endif
 #endif /* CONFIG_BOARD_THINGY91_NRF9160NS */
 
+#ifdef CONFIG_BOARD_CIRCUITDOJO_FEATHER_NRF9160NS
+#define AT_COEX0	"AT%XCOEX0=1,1,1565,1586"
```
Comment (tokangas) [inline samples/nrf9160/gps/src/main.c:44]: The percent sign could be escaped to be consistent with the other AT command string defines.
```suggestion
#define AT_COEX0	"AT\%XCOEX0=1,1,1565,1586"
```

**Review by tokangas — APPROVED** [review]
Look good.

File: samples/nrf9160/gps/src/main.c (hunk 1 above)
Comment (mniestroj) [reply samples/nrf9160/gps/src/main.c]: Done

**Comment by jtguggedal** [comment]
I think I'd prefer if we could make this generic instead of specific to the one board. How about introducing a couple of new Kconfigs in the sample Kconfig, so that we can do something liek this:

```
#if defined(CONFIG_GPS_SAMPLE_COEX0_CUSTOM)
#define AT_COEX0	CONFIG_GPS_SAMPLE_COEX0_STRING
#endif

#if defined(CONFIG_GPS_SAMPLE_MAGPIO_CUSTOM)
#define AT_MAGPIO	CONFIG_GPS_SAMPLE_MAGPIO_STRING
#endif

static const char *const at_commands[] = {
#if !defined(CONFIG_SUPL_CLIENT_LIB)
	AT_XSYSTEMMODE,
#endif
#ifdef AT_MAGPIO
	AT_MAGPIO,
#endif
#ifdef AT_COEX0
	AT_COEX0,
#endif
	AT_ACTIVATE_GPS
};
```

**Comment by mniestroj** [comment]
>I think I'd prefer if we could make this generic instead of specific to the one board. How about introducing a couple of new Kconfigs in the sample Kconfig, so that we can do something liek this:

I was thinking about something similar initially. I have also noticed that there is similar concept in the `drivers/gps/nrf9160_gps`, i.e. there are separate Kconfig options for AT_MAPGPIO and AT_COEX0. What about going one step further and converting `samples/nrf9160/gps` to use `drivers/gps/nrf9160_gps`? Not saying about doing it in this PR, but maybe it is the road to go directly instead of creating sample specific Kconfig options?

**Comment by tokangas** [comment]
> I was thinking about something similar initially. I have also noticed that there is similar concept in the `drivers/gps/nrf9160_gps`, i.e. there are separate Kconfig options for AT_MAPGPIO and AT_COEX0. What about going one step further and converting `samples/nrf9160/gps` to use `drivers/gps/nrf9160_gps`? Not saying about doing it in this PR, but maybe it is the road to go directly instead of creating sample specific Kconfig options?

GPS sample was updated to use the new GNSS API introduced in NCS v1.6.0. There are features, which are supported in the new API, but not in GPS driver. I'm not sure what are the actual future plans for the GPS driver, but personally I would prefer to go with the new GNSS API. The new API offers a simpler function based interface, so there may not be need for the GPS driver anymore.

**Comment by mniestroj** [comment]
I've updated this PR with introduction of sample specific Kconfig options as requested by @jtguggedal. Adding new board support still requires some board specific changes, but at least we don't bloat C code anymore and leave everything to Kconfig.

File: samples/nrf9160/gps/Kconfig [hunk 2]
```diff
@@ -17,12 +17,32 @@ choice GPS_SAMPLE_ANTENNA
...
 	bool "Use onboard antenna"
+	depends on !BOARD_CIRCUITDOJO_FEATHER_NRF9160NS
 
 config GPS_SAMPLE_ANTENNA_EXTERNAL
 	bool "Use external antenna"
 
 endchoice
 
+config GPS_SAMPLE_AT_MAGPIO
+	string "AT%MAGPIO command"
```
Comment (tokangas) [inline samples/nrf9160/gps/Kconfig:28]:
```suggestion
	string "AT%XMAGPIO command"
```

File: samples/nrf9160/gps/Kconfig [hunk 3]
```diff
@@ -17,12 +17,32 @@ choice GPS_SAMPLE_ANTENNA
...
+	default "AT\%XMAGPIO=1,0,0,1,1,1574,1577" if BOARD_NRF9160DK_NRF9160NS
+	default "AT\%XMAGPIO=1,1,1,7,1,746,803,2,698,748,2,1710,2200,3,824,894,4,880,960,5,791,849,7,1565,1586" if BOARD_THINGY91_NRF9160NS
+	help
+	  Defines what is the AT%MAGPIO command to be sent to GPS module. Leave
+	  empty if this commands should not be sent.
```
Comment (tokangas) [inline samples/nrf9160/gps/Kconfig:33]:
```suggestion
	  Defines what is the AT%XMAGPIO command to be sent to GPS module. Leave
	  empty if this command should not be sent.
```

File: samples/nrf9160/gps/Kconfig [hunk 4]
```diff
@@ -17,12 +17,32 @@ choice GPS_SAMPLE_ANTENNA
...
+
+config GPS_SAMPLE_AT_COEX0
+	string "AT%COEX0 command"
```
Comment (tokangas) [inline samples/nrf9160/gps/Kconfig:36]:
```suggestion
	string "AT%XCOEX0 command"
```

File: samples/nrf9160/gps/Kconfig [hunk 5]
```diff
@@ -17,12 +17,32 @@ choice GPS_SAMPLE_ANTENNA
...
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_NRF9160DK_NRF9160NS && GPS_SAMPLE_ANTENNA_ONBOARD)
+	default "AT\%XCOEX0" if (BOARD_NRF9160DK_NRF9160NS && GPS_SAMPLE_ANTENNA_EXTERNAL)
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_THINGY91_NRF9160NS && GPS_SAMPLE_ANTENNA_ONBOARD)
+	default "AT\%XCOEX0" if (BOARD_THINGY91_NRF9160NS && GPS_SAMPLE_ANTENNA_EXTERNAL)
+	default "AT\%XCOEX0=1,1,1565,1586" if BOARD_CIRCUITDOJO_FEATHER_NRF9160NS
+	help
+	  Defines what is the AT%COEX0 command to be sent to GPS module. Leave
+	  empty if this commands should not be sent.
```
Comment (tokangas) [inline samples/nrf9160/gps/Kconfig:44]:
```suggestion
	  Defines what is the AT%XCOEX0 command to be sent to GPS module. Leave
	  empty if this command should not be sent.
```

File: samples/nrf9160/gps/Kconfig [hunk 6]
```diff
@@ -17,12 +17,32 @@ choice GPS_SAMPLE_ANTENNA
...
+	help
+	  Defines what is the AT%MAGPIO command to be sent to GPS module. Leave
+	  empty if this commands should not be sent.
+
+config GPS_SAMPLE_AT_COEX0
+	string "AT%COEX0 command"
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_NRF9160DK_NRF9160NS && GPS_SAMPLE_ANTENNA_ONBOARD)
+	default "AT\%XCOEX0" if (BOARD_NRF9160DK_NRF9160NS && GPS_SAMPLE_ANTENNA_EXTERNAL)
+	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_THINGY91_NRF9160NS && GPS_SAMPLE_ANTENNA_ONBOARD)
+	default "AT\%XCOEX0" if (BOARD_THINGY91_NRF9160NS && GPS_SAMPLE_ANTENNA_EXTERNAL)
```
Comment (tokangas) [inline samples/nrf9160/gps/Kconfig:40]: DK and Thingy configuration could be combined in the same way as in the GPS driver Kconfig:
```suggestion
	default "AT\%XCOEX0=1,1,1565,1586" if (BOARD_NRF9160DK_NRF9160NS || BOARD_THINGY91_NRF9160NS) && GPS_SAMPLE_ANTENNA_ONBOARD
	default "AT\%XCOEX0" if (BOARD_NRF9160DK_NRF9160NS || BOARD_THINGY91_NRF9160NS) && GPS_SAMPLE_ANTENNA_EXTERNAL
```

**Review by tokangas — COMMENTED** [review]
Looks good to me. A few minor changes and the DK and Thingy lines in the Kconfig could be combined. @jtguggedal, what do you think?

File: samples/nrf9160/gps/Kconfig [hunk 7]
```diff
@@ -17,6 +17,7 @@ choice GPS_SAMPLE_ANTENNA
 
 config GPS_SAMPLE_ANTENNA_ONBOARD
 	bool "Use onboard antenna"
+	depends on !BOARD_CIRCUITDOJO_FEATHER_NRF9160NS
```
Comment (rlubos) [inline samples/nrf9160/gps/Kconfig:20]: @mniestroj We had an internal discussion regarding this and we've decided that we should not reference 3rd party boards in the NCS codebase - this simply doesn't scale for us. We don't have the means to verify that samples are operational on 3rd party devices on each release, and having them mentioned in the Kconfig files would suggest so.

Could you drop the commit adding the `BOARD_CIRCUITDOJO_FEATHER_NRF9160NS` support then? With other changes in this PR, it should be pretty straightforward to configure the sample for custom boards with only configuration changes.

Comment (mniestroj) [reply samples/nrf9160/gps/Kconfig:20]: >We had an internal discussion regarding this and we've decided that we should not reference 3rd party boards in the NCS codebase - this simply doesn't scale for us. We don't have the means to verify that samples are operational on 3rd party devices on each release, and having them mentioned in the Kconfig files would suggest so.

I understand that you do not have means for maintaining 3rd party boards. The alternative is to copy whole sample and adjust it for 3rd party boards. This is exactly what happens in https://github.com/circuitdojo/nrf9160-feather-examples-and-drivers/tree/v1.5.x/samples/gps. The result is that sdk-nrf's sample is compatible with nRF9160 DK and nrF9160 Thingy, while Circuitdojo's copy is an outdated GPS example which additionally supports nRF9160 Feather. Unfortunately this approach doesn't scale in another way, resulting in one repository per one board vendor. I think it is less work to maintain all boards in single repository, which is exactly what happens in mainline Zephyr. As nRF9160 GPS and LTE is supported only here, then in my opinion this is the repository in which nRF9160 boards should be supported. This gives users (and potential Nordic customers) one entry point for trying all boards. The documentation can simply state that Nordic is responsible for nRF9160 DK and Thingy, while other boards are supported by community.

>Could you drop the commit adding the BOARD_CIRCUITDOJO_FEATHER_NRF9160NS support then?

Created new #5079 PR, as the subject of intended changes is much different from original subject of this PR.

>With other changes in this PR, it should be pretty straightforward to configure the sample for custom boards with only configuration changes, that could be applied locally.

Pretty straighforward if we know what to set and to which values. Question is where should that information be documented and maintained? As written above, currently each vendor is responsible for creating new repository with copy-pasted samples and putting changes there. Alternatively board vendor can document what to configure before using samples from sdk-nrf, which complicates things for new users.

**Pull request closed without merging**